src/functional/ffront/decorator.py

```diff
@@ -167,62 +163,58 @@ def from_function(
+        source_def = SourceDefinition.from_function(definition)
+        closure_vars = {**(externals or {}), **get_closure_vars_from_function(definition)}
+        annotations = typing.get_type_hints(definition)
+        past_node = ProgramParser.apply(source_def, closure_vars, annotations)
         return cls(
             past_node=past_node,
-            captured_vars=captured_vars,
-            externals={} if externals is None else externals,
+            closure_vars=closure_vars,
             backend=backend,
             definition=definition,
             grid_type=grid_type,
         )
 
     def __post_init__(self):
-        # validate contents of captured vars
-        for name, value in self._filter_capture_vars_by_type(GTCallable).items():
-            if value.__gt_itir__().id != name:
-                raise RuntimeError(
-                    "Name of the closure reference and the function it holds do not match."
-                )
+        closure_var_functions = _filter_closure_vars_by_type(self.closure_vars, GTCallable)
+        if any(name != func.__gt_itir__().id for name, func in closure_var_functions.items()):
+            raise RuntimeError(
+                "The symbol name and the name of the function identified by the symbol are not equal."
+            )
 
-        # validate Symbols of captured vars in PAST
-        referenced_var_names: set[str] = set()
-        for captured_var in self.past_node.captured_vars:
-            if isinstance(captured_var.type, (ct.CallableType, ct.OffsetType, ct.DimensionType)):
-                referenced_var_names.add(captured_var.id)
-            else:
-                raise NotImplementedError("Only function closure vars are allowed currently.")
-        defined_var_names = set(self.all_capture_vars.globals) | set(
-            self.all_capture_vars.nonlocals
-        )
-        if undefined := referenced_var_names - defined_var_names:
-            raise RuntimeError(f"Reference to undefined symbol(s) `{', '.join(undefined)}`.")
+        undefined_symbols = [
+            symbol.id
+            for symbol in self.past_node.closure_symbols
+            if symbol.id not in self.closure_vars
+        ]
+        if undefined_symbols:
+            raise RuntimeError(
+                f"The following closure variables are undefined: {', '.join(undefined_symbols)}"
+            )
 
     def with_backend(self, backend: FencilExecutor) -> "Program":
         return Program(
             past_node=self.past_node,
-            captured_vars=self.captured_vars,
-            externals=self.externals,
+            closure_vars=self.closure_vars,
             backend=backend,
             definition=self.definition,  # type: ignore[arg-type]  # mypy wrongly deduces definition as method here
         )
 
     @functools.cached_property
-    def all_capture_vars(self) -> CapturedVars:
-        return _collect_capture_vars(self.captured_vars)
+    def closure_vars_recursive(self) -> dict[str, Any]:
```

Comment:
@tehrengruber doesn't like `recursive_closure_vars` because the closure variables themselves are not recursive, they are just recursively collected. I don't like `all_closure_vars` because *all* is not very informative.

Otherwise, I agree it should be renamed. Do you have any other suggestions, @egparedes ?

Reply:
Solved.